app/models/creatorship.rb

```diff
@@ -1,47 +1,276 @@
+
+  ########################################
+  # VALIDATIONS
+  ########################################
+
+  before_validation :update_approved, on: :create
+
+  validates_presence_of :creation
+  validates_uniqueness_of :pseud, scope: [:creation_type, :creation_id], on: :create
+
+  validate :check_invalid, on: :create
+  validate :check_disallowed, on: :create
+  validate :check_banned, on: :create
+  validate :check_approved_becoming_false, on: :update
+
+  # Update approval status if this creatorship should be automatically approved.
+  def update_approved
+    if !approved? && should_automatically_approve?
+      self.approved = true
+    end
+  end
+
+  # Make sure that the pseud exists, and isn't ambiguous.
+  #
+  # Note that thanks to the definitions of missing? and ambiguous?, this is
+  # equivalent to having a validates_presence_of :pseud check, just with
+  # different error messages.
+  def check_invalid
+    if missing?
+      errors.add(:base, ts("Could not find a pseud %{name}.", name: byline))
+    elsif ambiguous?
+      errors.add(:base, ts("The pseud %{name} is ambiguous.", name: byline))
+    end
+  end
+
+  # Make sure that if this is an invitation, we're not inviting someone who has
+  # disabled invitations.
+  def check_disallowed
+    return if approved? || pseud.nil?
+    return if pseud&.user&.preference&.allow_cocreator
+
+    errors.add(:base, ts("%{name} does not allow others to add them as a co-creator.",
+                         name: pseud.byline))
+  end
+
+  # Make sure that the user isn't banned or suspended.
+  def check_banned
+    return unless pseud&.user&.banned || pseud&.user&.suspended
+
+    errors.add(:base, ts("%{name} is currently banned and cannot be listed as a co-creator.",
+                         name: pseud.byline))
+  end
+
+  # Make sure that we're not trying to set approved to false, since that could
+  # potentially violate some rules about co-creators. (e.g. Having a user
+  # listed as a chapter co-creator, but not a work co-creator.)
+  def check_approved_becoming_false
+    if !approved? && approved_changed?
+      errors.add(:base, "Once approved, a creatorship cannot become unapproved.")
+    end
+  end
+
+  ########################################
+  # CALLBACKS
+  ########################################
+
+  after_create :add_to_parents
+  after_update :add_to_parents, if: :saved_change_to_approved?
 
   before_destroy :expire_caches
-  after_create :update_pseud_index
-  after_destroy :update_pseud_index
+  before_destroy :check_not_last
+  after_destroy :remove_from_children
+
+  after_commit :update_indices
+
+  after_create_commit :notify_creator, if: :enable_notifications
+
+  # If a pseud is listed as a work co-creator (not invited, actually listed),
+  # they should also be listed on all of the work's series. Similarly, if a
+  # pseud is listed as a chapter co-creator, they should also be listed on the
+  # work.
+  def add_to_parents
```

Comment:
We don't remove from parents right? e.g. if a co-creator is removed from all chapters / works in a series, they should still be a co-creator in the work / series?

Reply:
Right.